boards/DIM/Test/Src/seven_seg_displays_test.cpp

```diff
@@ -0,0 +1,118 @@
+
+    App_SevenSegDisplays_WriteStateOfCharge(seven_segment_displays);
+
+    ASSERT_EQ(1, set_left_hex_digit_fake.arg0_history[0]);
+    ASSERT_EQ(0, set_middle_hex_digit_fake.arg0_history[0]);
+    ASSERT_EQ(0, set_right_hex_digit_fake.arg0_history[0]);
+
+    ASSERT_EQ(1, get_state_of_charge_fake.call_count);
+    ASSERT_EQ(1, set_left_hex_digit_fake.call_count);
+    ASSERT_EQ(1, set_middle_hex_digit_fake.call_count);
+    ASSERT_EQ(1, set_right_hex_digit_fake.call_count);
+
+    ASSERT_EQ(fff.call_history[0], (void *)get_state_of_charge);
+    ASSERT_EQ(fff.call_history[1], (void *)set_left_hex_digit);
+    ASSERT_EQ(fff.call_history[2], (void *)set_middle_hex_digit);
+    ASSERT_EQ(fff.call_history[3], (void *)set_right_hex_digit);
+}
+
+TEST_F(SevenSegDisplaysTest, write_min_state_of_chrage)
+{
+    get_state_of_charge_fake.return_val = 0.0f;
+
+    App_SevenSegDisplays_WriteStateOfCharge(seven_segment_displays);
+
+    ASSERT_EQ(0, set_left_hex_digit_fake.arg0_history[0]);
+    ASSERT_EQ(0, set_middle_hex_digit_fake.arg0_history[0]);
+    ASSERT_EQ(0, set_right_hex_digit_fake.arg0_history[0]);
+
+    ASSERT_EQ(1, get_state_of_charge_fake.call_count);
+    ASSERT_EQ(1, set_left_hex_digit_fake.call_count);
+    ASSERT_EQ(1, set_middle_hex_digit_fake.call_count);
+    ASSERT_EQ(1, set_right_hex_digit_fake.call_count);
+
+    ASSERT_EQ(fff.call_history[0], (void *)get_state_of_charge);
+    ASSERT_EQ(fff.call_history[1], (void *)set_left_hex_digit);
+    ASSERT_EQ(fff.call_history[2], (void *)set_middle_hex_digit);
+    ASSERT_EQ(fff.call_history[3], (void *)set_right_hex_digit);
+}
+
+TEST_F(SevenSegDisplaysTest, write_state_of_chrage_with_decimals)
+{
+    get_state_of_charge_fake.return_val = 23.456789f;
+
+    App_SevenSegDisplays_WriteStateOfCharge(seven_segment_displays);
+
+    ASSERT_EQ(0, set_left_hex_digit_fake.arg0_history[0]);
+    ASSERT_EQ(2, set_middle_hex_digit_fake.arg0_history[0]);
+    ASSERT_EQ(3, set_right_hex_digit_fake.arg0_history[0]);
+
+    ASSERT_EQ(1, get_state_of_charge_fake.call_count);
+    ASSERT_EQ(1, set_left_hex_digit_fake.call_count);
+    ASSERT_EQ(1, set_middle_hex_digit_fake.call_count);
+    ASSERT_EQ(1, set_right_hex_digit_fake.call_count);
+
+    ASSERT_EQ(fff.call_history[0], (void *)get_state_of_charge);
+    ASSERT_EQ(fff.call_history[1], (void *)set_left_hex_digit);
+    ASSERT_EQ(fff.call_history[2], (void *)set_middle_hex_digit);
+    ASSERT_EQ(fff.call_history[3], (void *)set_right_hex_digit);
+}
+
+TEST_F(SevenSegDisplaysTest, write_overflow_state_of_charge)
+{
+    get_state_of_charge_fake.return_val = 101.0f;
+
+    ASSERT_DEATH(
+        { App_SevenSegDisplays_WriteStateOfCharge(seven_segment_displays); },
+        "");
+}
+
+TEST_F(SevenSegDisplaysTest, write_underflow_state_of_charge)
+{
+    get_state_of_charge_fake.return_val = -1.0f;
+
+    ASSERT_DEATH(
+        { App_SevenSegDisplays_WriteStateOfCharge(seven_segment_displays); },
+        "");
```

Comment:
Do we really want to hardfault if we try to write an invalid SoC to the seven seg? It seems to me that a more reasonable course for something non-critical like this would be to return a value that we can check, and then log a CAN message if the returned value indicates the sevenseg could not write what we asked it to.

Reply:
I think you're right on this one and it really speaks on how we avoid error codes like the plague in our codebase. Something I've really wanted to start doing is introducing generic error codes that can be shared across the modules.

Something like
```
// StatusCodes for various errors, keep these generic. Never assume their order is fixed so refer to
// them by name only.
typedef enum {
  STATUS_CODE_OK = 0,
  STATUS_CODE_UNKNOWN,
  STATUS_CODE_INVALID_ARGS,
  STATUS_CODE_RESOURCE_EXHAUSTED,
  STATUS_CODE_UNREACHABLE,
  STATUS_CODE_TIMEOUT,
  STATUS_CODE_EMPTY,
  STATUS_CODE_OUT_OF_RANGE,
  STATUS_CODE_UNIMPLEMENTED,
  STATUS_CODE_UNINITIALIZED,
  STATUS_CODE_INTERNAL_ERROR,
  NUM_STATUS_CODE
```
And then we can `ASSERT_EQ` on return codes rather than `DEATH`. What do you think? If you think it's sane I would like to add `error_code.h` in this PR.

Reply:
Sure, that's fine by me, this is essentially our equivalent of `HAL_Status`

Reply:
Added some error codes under `App_ErrorCode.h`. Later I'd like to add some simple macros to log some extra information (e.g. `__LINE__`, custom message describing the error, etc) but that's for another PR